common/scala/src/main/scala/org/apache/openwhisk/core/yarn/YARNContainerFactory.scala

```diff
@@ -0,0 +1,243 @@
+                           containerArgs: ContainerArgsConfig =
+                             loadConfigOrThrow[ContainerArgsConfig](ConfigKeys.containerArgs),
+                           yarnConfig: YARNConfig = loadConfigOrThrow[YARNConfig](ConfigKeys.yarn))
+    extends ContainerFactory {
+
+  val images: Set[ImageName] = ExecManifest.runtimesManifest.runtimes.flatMap(a => a.versions.map(b => b.image))
+
+  //One actor of each type per image for parallelism
+  private var yarnComponentActors: Map[ImageName, ActorRef] = HashMap[ImageName, ActorRef]()
+  private var YARNContainerInfoActors: Map[ImageName, ActorRef] = HashMap[ImageName, ActorRef]()
+
+  val serviceStartTimeoutMS = 60000
+  val retryWaitMS = 1000
+  val runCommand = ""
+  val version = "1.0.0"
+  val description = "OpenWhisk Action Service"
+
+  //Allows for invoker HA
+  val serviceName: String = yarnConfig.serviceName + "-" + instance.toInt
+
+  val containerStartTimeoutMS = 60000
+
+  implicit val as: ActorSystem = actorSystem
+  implicit val materializer: ActorMaterializer = ActorMaterializer()
+  implicit val ec: ExecutionContext = actorSystem.dispatcher
+
+  override def init(): Unit = {
+    yarnComponentActors = images
+      .map(
+        i =>
+          (
+            i,
+            actorSystem.actorOf(
+              Props(new YARNComponentActor(actorSystem, logging, yarnConfig, serviceName, i)),
+              name = s"YARNComponentActor-${i.name}")))
+      .toMap
+    YARNContainerInfoActors = images
+      .map(
+        i =>
+          (
+            i,
+            actorSystem.actorOf(
+              Props(new YARNContainerInfoActor(actorSystem, logging, yarnConfig, serviceName, i)),
+              name = s"YARNComponentInfoActor-${i.name}")))
+      .toMap
+    blocking {
+      implicit val timeout: Timeout = Timeout(serviceStartTimeoutMS.milliseconds)
+
+      //Remove service if it already exists
+      val serviceDef =
+        YARNRESTUtil.downloadServiceDefinition(yarnConfig.authType, serviceName, yarnConfig.masterUrl)(logging)
+
+      if (serviceDef != null)
+        removeService()
+
+      createService()
+    }
+  }
+  override def createContainer(
+    unusedtid: TransactionId,
+    unusedname: String,
+    actionImage: ExecManifest.ImageName,
+    unuseduserProvidedImage: Boolean,
+    unusedmemory: ByteSize,
+    unusedcpuShares: Int)(implicit config: WhiskConfig, logging: Logging): Future[Container] = {
+    implicit val timeout: Timeout = Timeout(containerStartTimeoutMS.milliseconds)
+
+    //First send the create command to YARN, then with a different actor, wait for the container to be ready
+    ask(yarnComponentActors(actionImage), CreateContainerAsync).flatMap(_ =>
+      ask(YARNContainerInfoActors(actionImage), GetContainerInfo(yarnComponentActors(actionImage))).mapTo[Container])
+  }
+  override def cleanup(): Unit = {
+    removeService()
+    yarnComponentActors.foreach({ case (k, v)     => actorSystem.stop(v) })
```

Comment:
you dont need the outer `()` foreach: `foreach { case (k, v)     => actorSystem.stop(v) }` works.